In [2]:
from numpy import array
length = 5
seq = array([i/float(length) for i in range(length)])
print(seq)

[0.  0.2 0.4 0.6 0.8]


In [3]:
X = seq.reshape(5, 1, 1)
y = seq.reshape(5, 1)

In [4]:
X

array([[[0. ]],

       [[0.2]],

       [[0.4]],

       [[0.6]],

       [[0.8]]])

## One-to-One LSTM for Sequence Prediction

In [5]:
from numpy import array
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
# prepare sequence
length = 5
seq = array([i/float(length) for i in range(length)])
X = seq.reshape(len(seq), 1, 1)
y = seq.reshape(len(seq), 1)
# define LSTM configuration
n_neurons = length
n_batch = length
n_epoch = 1000
# create LSTM
model = Sequential()
model.add(LSTM(n_neurons, input_shape=(1, 1)))
model.add(Dense(1))
model.compile(loss='mean_squared_error', optimizer='adam')
print(model.summary())
# train LSTM
model.fit(X, y, epochs=n_epoch, batch_size=n_batch, verbose=2)
# evaluate
result = model.predict(X, batch_size=n_batch, verbose=0)
for value in result:
    print('%.1f' % value)

Using TensorFlow backend.


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
lstm_1 (LSTM)                (None, 5)                 140       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 6         
Total params: 146
Trainable params: 146
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/1000
 - 4s - loss: 0.3464
Epoch 2/1000
 - 0s - loss: 0.3448
Epoch 3/1000
 - 0s - loss: 0.3429
Epoch 4/1000
 - 0s - loss: 0.3409
Epoch 5/1000
 - 0s - loss: 0.3389
Epoch 6/1000
 - 0s - loss: 0.3368
Epoch 7/1000
 - 0s - loss: 0.3347
Epoch 8/1000
 - 0s - loss: 0.3326
Epoch 9/1000
 - 0s - loss: 0.3306
Epoch 10/1000
 - 0s - loss: 0.3285
Epoch 11/1000
 - 0s - loss: 0.3264
Epoch 12/1000
 - 0s - loss: 0.3243
Epoch 13/1000
 - 0s - loss: 0.3223
Epoch 14/1000
 - 0s - loss: 0.3202
Epoch 15/1000
 - 0s - loss: 0.3182
Epoch 16/1000
 - 0s

Epoch 215/1000
 - 0s - loss: 0.1032
Epoch 216/1000
 - 0s - loss: 0.1027
Epoch 217/1000
 - 0s - loss: 0.1022
Epoch 218/1000
 - 0s - loss: 0.1017
Epoch 219/1000
 - 0s - loss: 0.1012
Epoch 220/1000
 - 0s - loss: 0.1007
Epoch 221/1000
 - 0s - loss: 0.1003
Epoch 222/1000
 - 0s - loss: 0.0998
Epoch 223/1000
 - 0s - loss: 0.0993
Epoch 224/1000
 - 0s - loss: 0.0988
Epoch 225/1000
 - 0s - loss: 0.0984
Epoch 226/1000
 - 0s - loss: 0.0979
Epoch 227/1000
 - 0s - loss: 0.0975
Epoch 228/1000
 - 0s - loss: 0.0970
Epoch 229/1000
 - 0s - loss: 0.0965
Epoch 230/1000
 - 0s - loss: 0.0961
Epoch 231/1000
 - 0s - loss: 0.0957
Epoch 232/1000
 - 0s - loss: 0.0952
Epoch 233/1000
 - 0s - loss: 0.0948
Epoch 234/1000
 - 0s - loss: 0.0943
Epoch 235/1000
 - 0s - loss: 0.0939
Epoch 236/1000
 - 0s - loss: 0.0935
Epoch 237/1000
 - 0s - loss: 0.0931
Epoch 238/1000
 - 0s - loss: 0.0926
Epoch 239/1000
 - 0s - loss: 0.0922
Epoch 240/1000
 - 0s - loss: 0.0918
Epoch 241/1000
 - 0s - loss: 0.0914
Epoch 242/1000
 - 0s - loss:

Epoch 443/1000
 - 0s - loss: 0.0494
Epoch 444/1000
 - 0s - loss: 0.0492
Epoch 445/1000
 - 0s - loss: 0.0491
Epoch 446/1000
 - 0s - loss: 0.0489
Epoch 447/1000
 - 0s - loss: 0.0488
Epoch 448/1000
 - 0s - loss: 0.0486
Epoch 449/1000
 - 0s - loss: 0.0485
Epoch 450/1000
 - 0s - loss: 0.0483
Epoch 451/1000
 - 0s - loss: 0.0482
Epoch 452/1000
 - 0s - loss: 0.0480
Epoch 453/1000
 - 0s - loss: 0.0479
Epoch 454/1000
 - 0s - loss: 0.0478
Epoch 455/1000
 - 0s - loss: 0.0476
Epoch 456/1000
 - 0s - loss: 0.0475
Epoch 457/1000
 - 0s - loss: 0.0473
Epoch 458/1000
 - 0s - loss: 0.0472
Epoch 459/1000
 - 0s - loss: 0.0470
Epoch 460/1000
 - 0s - loss: 0.0469
Epoch 461/1000
 - 0s - loss: 0.0467
Epoch 462/1000
 - 0s - loss: 0.0466
Epoch 463/1000
 - 0s - loss: 0.0464
Epoch 464/1000
 - 0s - loss: 0.0463
Epoch 465/1000
 - 0s - loss: 0.0461
Epoch 466/1000
 - 0s - loss: 0.0460
Epoch 467/1000
 - 0s - loss: 0.0458
Epoch 468/1000
 - 0s - loss: 0.0457
Epoch 469/1000
 - 0s - loss: 0.0455
Epoch 470/1000
 - 0s - loss:

Epoch 671/1000
 - 0s - loss: 0.0144
Epoch 672/1000
 - 0s - loss: 0.0143
Epoch 673/1000
 - 0s - loss: 0.0142
Epoch 674/1000
 - 0s - loss: 0.0140
Epoch 675/1000
 - 0s - loss: 0.0139
Epoch 676/1000
 - 0s - loss: 0.0138
Epoch 677/1000
 - 0s - loss: 0.0137
Epoch 678/1000
 - 0s - loss: 0.0135
Epoch 679/1000
 - 0s - loss: 0.0134
Epoch 680/1000
 - 0s - loss: 0.0133
Epoch 681/1000
 - 0s - loss: 0.0132
Epoch 682/1000
 - 0s - loss: 0.0130
Epoch 683/1000
 - 0s - loss: 0.0129
Epoch 684/1000
 - 0s - loss: 0.0128
Epoch 685/1000
 - 0s - loss: 0.0127
Epoch 686/1000
 - 0s - loss: 0.0126
Epoch 687/1000
 - 0s - loss: 0.0125
Epoch 688/1000
 - 0s - loss: 0.0123
Epoch 689/1000
 - 0s - loss: 0.0122
Epoch 690/1000
 - 0s - loss: 0.0121
Epoch 691/1000
 - 0s - loss: 0.0120
Epoch 692/1000
 - 0s - loss: 0.0119
Epoch 693/1000
 - 0s - loss: 0.0118
Epoch 694/1000
 - 0s - loss: 0.0116
Epoch 695/1000
 - 0s - loss: 0.0115
Epoch 696/1000
 - 0s - loss: 0.0114
Epoch 697/1000
 - 0s - loss: 0.0113
Epoch 698/1000
 - 0s - loss:

Epoch 898/1000
 - 0s - loss: 8.9663e-04
Epoch 899/1000
 - 0s - loss: 8.8552e-04
Epoch 900/1000
 - 0s - loss: 8.7460e-04
Epoch 901/1000
 - 0s - loss: 8.6385e-04
Epoch 902/1000
 - 0s - loss: 8.5329e-04
Epoch 903/1000
 - 0s - loss: 8.4291e-04
Epoch 904/1000
 - 0s - loss: 8.3270e-04
Epoch 905/1000
 - 0s - loss: 8.2266e-04
Epoch 906/1000
 - 0s - loss: 8.1280e-04
Epoch 907/1000
 - 0s - loss: 8.0310e-04
Epoch 908/1000
 - 0s - loss: 7.9356e-04
Epoch 909/1000
 - 0s - loss: 7.8419e-04
Epoch 910/1000
 - 0s - loss: 7.7498e-04
Epoch 911/1000
 - 0s - loss: 7.6593e-04
Epoch 912/1000
 - 0s - loss: 7.5704e-04
Epoch 913/1000
 - 0s - loss: 7.4829e-04
Epoch 914/1000
 - 0s - loss: 7.3970e-04
Epoch 915/1000
 - 0s - loss: 7.3126e-04
Epoch 916/1000
 - 0s - loss: 7.2296e-04
Epoch 917/1000
 - 0s - loss: 7.1481e-04
Epoch 918/1000
 - 0s - loss: 7.0680e-04
Epoch 919/1000
 - 0s - loss: 6.9893e-04
Epoch 920/1000
 - 0s - loss: 6.9120e-04
Epoch 921/1000
 - 0s - loss: 6.8360e-04
Epoch 922/1000
 - 0s - loss: 6.7614e-04


We can see that the LSTM layer has 140 parameters. This is calculated based on the number of inputs (1) and the number of outputs (5 for the 5 units in the hidden layer), as follows:  
n = 4 * ((inputs + 1) * outputs + outputs^2)  
n = 4 * ((1 + 1) * 5 + 5^2)  
n = 4 * 35  
n = 140  

We can also see that the fully connected layer only has 6 parameters for the number of inputs (5 for the 5 inputs from the previous layer), number of outputs (1 for the 1 neuron in the layer), and the bias.  
n = inputs * outputs + outputs  
n = 5 * 1 + 1  
n = 6  

## Many-to-One LSTM for Sequence Prediction (without  TimeDistributed)
In this section, we develop an LSTM to output the sequence all at once, although without the TimeDistributed wrapper layer.

The input for LSTMs must be three dimensional. We can reshape the 2D sequence into a 3D sequence with 1 sample, 5 time steps, and 1 feature. We will define the output as 1 sample with 5 features.

In [6]:
X = seq.reshape(1, 5, 1)
y = seq.reshape(1, 5)

Immediately, you can see that the problem definition must be slightly adjusted to support a network for sequence prediction without a TimeDistributed wrapper. Specifically, output one vector rather build out an output sequence one step at a time. The difference may sound subtle, but it is important to understanding the role of the TimeDistributed wrapper.

We will define the model as having one input with 5 time steps. The first hidden layer will be an LSTM with 5 units. The output layer is a fully-connected layer with 5 neurons.

In [8]:
from numpy import array
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
# prepare sequence
length = 5
seq = array([i/float(length) for i in range(length)])
X = seq.reshape(1, length, 1)
y = seq.reshape(1, length)
# define LSTM configuration
n_neurons = length
n_batch = 1
n_epoch = 500
# create LSTM
model = Sequential()
model.add(LSTM(n_neurons, input_shape=(length, 1)))
model.add(Dense(length))
model.compile(loss='mean_squared_error', optimizer='adam')
print(model.summary())
# train LSTM
model.fit(X, y, epochs=n_epoch, batch_size=n_batch, verbose=2)
# evaluate
result = model.predict(X, batch_size=n_batch, verbose=0)
for value in result[0,:]:
    print('%.1f' % value)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
lstm_3 (LSTM)                (None, 5)                 140       
_________________________________________________________________
dense_3 (Dense)              (None, 5)                 30        
Total params: 170
Trainable params: 170
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/500
 - 1s - loss: 0.3042
Epoch 2/500
 - 0s - loss: 0.3025
Epoch 3/500
 - 0s - loss: 0.3006
Epoch 4/500
 - 0s - loss: 0.2987
Epoch 5/500
 - 0s - loss: 0.2967
Epoch 6/500
 - 0s - loss: 0.2947
Epoch 7/500
 - 0s - loss: 0.2927
Epoch 8/500
 - 0s - loss: 0.2907
Epoch 9/500
 - 0s - loss: 0.2887
Epoch 10/500
 - 0s - loss: 0.2868
Epoch 11/500
 - 0s - loss: 0.2848
Epoch 12/500
 - 0s - loss: 0.2829
Epoch 13/500
 - 0s - loss: 0.2810
Epoch 14/500
 - 0s - loss: 0.2792
Epoch 15/500
 - 0s - loss: 0.2773
Epoch 16/500
 - 0s - loss: 0.2755


 - 0s - loss: 0.0143
Epoch 222/500
 - 0s - loss: 0.0136
Epoch 223/500
 - 0s - loss: 0.0128
Epoch 224/500
 - 0s - loss: 0.0121
Epoch 225/500
 - 0s - loss: 0.0115
Epoch 226/500
 - 0s - loss: 0.0108
Epoch 227/500
 - 0s - loss: 0.0102
Epoch 228/500
 - 0s - loss: 0.0096
Epoch 229/500
 - 0s - loss: 0.0091
Epoch 230/500
 - 0s - loss: 0.0085
Epoch 231/500
 - 0s - loss: 0.0080
Epoch 232/500
 - 0s - loss: 0.0075
Epoch 233/500
 - 0s - loss: 0.0070
Epoch 234/500
 - 0s - loss: 0.0066
Epoch 235/500
 - 0s - loss: 0.0061
Epoch 236/500
 - 0s - loss: 0.0057
Epoch 237/500
 - 0s - loss: 0.0054
Epoch 238/500
 - 0s - loss: 0.0050
Epoch 239/500
 - 0s - loss: 0.0046
Epoch 240/500
 - 0s - loss: 0.0043
Epoch 241/500
 - 0s - loss: 0.0040
Epoch 242/500
 - 0s - loss: 0.0037
Epoch 243/500
 - 0s - loss: 0.0034
Epoch 244/500
 - 0s - loss: 0.0032
Epoch 245/500
 - 0s - loss: 0.0030
Epoch 246/500
 - 0s - loss: 0.0027
Epoch 247/500
 - 0s - loss: 0.0025
Epoch 248/500
 - 0s - loss: 0.0023
Epoch 249/500
 - 0s - loss: 0.0021

 - 0s - loss: 8.4747e-12
Epoch 436/500
 - 0s - loss: 7.6057e-12
Epoch 437/500
 - 0s - loss: 7.1588e-12
Epoch 438/500
 - 0s - loss: 6.3798e-12
Epoch 439/500
 - 0s - loss: 5.6792e-12
Epoch 440/500
 - 0s - loss: 5.2285e-12
Epoch 441/500
 - 0s - loss: 4.7025e-12
Epoch 442/500
 - 0s - loss: 4.0483e-12
Epoch 443/500
 - 0s - loss: 3.4404e-12
Epoch 444/500
 - 0s - loss: 3.0996e-12
Epoch 445/500
 - 0s - loss: 2.7335e-12
Epoch 446/500
 - 0s - loss: 2.3739e-12
Epoch 447/500
 - 0s - loss: 1.9203e-12
Epoch 448/500
 - 0s - loss: 1.6506e-12
Epoch 449/500
 - 0s - loss: 1.4082e-12
Epoch 450/500
 - 0s - loss: 1.1788e-12
Epoch 451/500
 - 0s - loss: 1.0549e-12
Epoch 452/500
 - 0s - loss: 8.1332e-13
Epoch 453/500
 - 0s - loss: 6.7513e-13
Epoch 454/500
 - 0s - loss: 5.3779e-13
Epoch 455/500
 - 0s - loss: 4.5184e-13
Epoch 456/500
 - 0s - loss: 3.9170e-13
Epoch 457/500
 - 0s - loss: 3.4293e-13
Epoch 458/500
 - 0s - loss: 3.0131e-13
Epoch 459/500
 - 0s - loss: 2.8577e-13
Epoch 460/500
 - 0s - loss: 2.4399e-13


The LSTM units have been crippled and will each output a single value, providing a vector of 5 values as inputs to the fully connected layer. The time dimension or sequence information has been thrown away and collapsed into a vector of 5 values.

## Many-to-Many LSTM for Sequence Prediction (with TimeDistributed)